R/newton_solver.R

```diff
@@ -0,0 +1,80 @@
+#' Newton solver implemented in R
+#'
+#' @param contact_matrix Social contact matrix. Entry $mm_{ij}$ gives average
+#' number of contacts in group $i$ reported by participants in group j
+#' @param demography_vector Demography vector. Entry $pp_{i}$ gives proportion
+#' of total population in group $i$ (model will normalise if needed)
+#' @param susceptibility A matrix giving the susceptibility of individuals in
+#' demographic group $i$ and risk group $j$.
+#' @param iterations Number of solver iterations. Defaults to 1,000.
+#' @param tolerance Solver error tolerance.
+#'
+#' @return A vector final sizes, of the size (N demography groups *
+#' N risk groups).
+solve_final_size_newton <- function(contact_matrix,
+                                    demography_vector,
+                                    susceptibility,
+                                    iterations = 1000,
+                                    tolerance = 1e-6) {
+  # count demography groups
+  n_dim <- length(demography_vector)
+
+  # make vector of zeros
+  zeros <- rep(0.0, n_dim)
+  # make vector of initial final size guesses = 0.5
+  epi_final_size <- rep(0.5, n_dim)
+
+  # set contact_matrix values to zero if there are no contacts among
+  # demography groups, or if demography groups are empty
+  i_here <- demography_vector == 0 | susceptibility == 0 |
+    rowSums(contact_matrix) == 0
+  zeros[i_here] <- 1.0
+  epi_final_size[i_here] <- 0.0
+
+  # matrix filled by columns
+  contact_matrix_ <- contact_matrix * demography_vector %o% susceptibility
+
+  contact_matrix_[i_here, i_here] <- 0.0
+
+  # make a copy of the initial guesses for final size
+  cache_v <- epi_final_size
+
+  # define functions for matrix operations
+  f1 <- function(x) {
+    # contact_matrix_ captured from environment
+    (contact_matrix_ %*% (1.0 - x)) + log(x)
+  }
+
+  f2 <- function(x) {
+    # contact_matrix_ captured from environment
+    -contact_matrix_ + diag(1.0 / x)
+  }
+
+  # cache_m is the contact matrix. must not be modified!
+  dx_f <- function(x) {
+    cache_m <- f2(x)
+    cache <- -f1(x)
+    # partial pivoting LU decomposition
+    cache_m_pivlu <- Matrix::lu(cache_m)
+    cache_m_pivlu <- Matrix::expand(cache_m_pivlu)
+    cache_m_pivlu <- cache_m_pivlu$L %*% cache_m_pivlu$U
+    # return solution
+    solve(cache_m_pivlu, cache)
+  }
+
+  # prepare an initial value from where to
+  x_ <- rep(1e-6, n_dim)
+
+  for (i in seq(iterations)) {
```

Comment:
Should there be a warning if we reach the maximum number of iterations but `error_` is still much larger than the `tolerance`?

Reply:
What should we consider much larger?

Reply:
I added an arbitary value of 100x tolerance (typically 1e-4 as default tolerance is 1e-6), as 10x was leading to a number of warnings when iterations < 1e4.

Reply:
Also added checks to expect warnings for both solvers